wormchain/contracts/tools/deploy_wormchain.ts

```diff
@@ -0,0 +1,201 @@
+  let msg = `.env is missing. run "make contracts-tools-deps" to fetch.`
+  console.error(msg)
+  throw msg
+}
+
+const readFileAsync = util.promisify(fs.readFile);
+
+/*
+  NOTE: Only append to this array: keeping the ordering is crucial, as the
+  contracts must be imported in a deterministic order so their addresses remain
+  deterministic.
+*/
+type ContractName = string
+const artifacts: ContractName[] = [
+  "cw20_base.wasm",
+  "wormchain_accounting.wasm",
+];
+
+const ARTIFACTS_PATH = "../artifacts/"
+/* Check that the artifact folder contains all the wasm files we expect and nothing else */
+
+const actual_artifacts = readdirSync(ARTIFACTS_PATH).filter((a) =>
+  a.endsWith(".wasm")
+);
+
+const missing_artifacts = artifacts.filter(
+  (a) => !actual_artifacts.includes(a)
+);
+if (missing_artifacts.length) {
+  console.log(
+    "Error during wormchain deployment. The following files are expected to be in the artifacts folder:"
+  );
+  missing_artifacts.forEach((file) => console.log(`  - ${file}`));
+  console.log(
+    "Hint: the deploy script needs to run after the contracts have been built."
+  );
+  console.log(
+    "External binary blobs need to be manually added in tools/Dockerfile."
+  );
+  process.exit(1);
+}
+
+
+
+async function main() {
+
+  /* Set up cosmos client & wallet */
+
+  let host = devnetConsts.chains[3104].tendermintUrlLocal
+  if (os.hostname().includes("wormchain-deploy")) {
+    // running in tilt devnet
+    host = devnetConsts.chains[3104].tendermintUrlTilt
+  }
+  const denom = devnetConsts.chains[3104].addresses.native.denom
+  const mnemonic = devnetConsts.chains[3104].accounts.wormchainNodeOfGuardian0.mnemonic
+  const addressPrefix = "wormhole"
+
+  const w = await Secp256k1HdWallet.fromMnemonic(mnemonic, { prefix: addressPrefix })
+
+
+  const gas = GasPrice.fromString(`0${denom}`)
+  let cwc: SigningCosmWasmClient
+  try {
+    cwc = await SigningCosmWasmClient.connectWithSigner(host, w, { prefix: addressPrefix, gasPrice: gas })
+  } catch (e) {
+    let msg = `could not connect to wormchain host: ${host}`
+    if (e?.message) {
+      console.error(e.message)
+    }
+    throw msg
+  }
+
+
+  // there are several Cosmos chains in devnet, so check the config is as expected
+  let id = await cwc.getChainId()
+  if (id !== "wormchain") {
+    throw new Error(`Wormchain CosmWasmClient connection produced an unexpected chainID: ${id}`)
+  }
+
+  const signers = await w.getAccounts()
+  const signer = signers[0].address
+  console.log("wormchain contract deployer is: ", signer)
+
+  /* Deploy artifacts */
+
+  const codeIds: { [name: ContractName]: number } = await artifacts.reduce(async (prev, file) => {
+    // wait for the previous to finish, to avoid the race condition of wallet sequence mismatch.
+    const accum = await prev
+
+    const contract_bytes = await readFileAsync(`${ARTIFACTS_PATH}${file}`);
+
+    const i = await cwc.upload(signer, contract_bytes, "auto", "")
```

Comment:
I'm surprised this works.  I thought we changed things so that contracts could only be uploaded via a governance action in #1674.

@conorpp thoughts?

Reply:
You are correct, the deploy permissions are not effective. Conor found the issue, we will implement permissioned deploy in the future, tracking with https://github.com/wormhole-foundation/wormhole/issues/2276.